### ISSUE TITLE
Using AsNoTracking() is slower over multiple iterations of a simple query

### ISSUE THREAD

**Issue by brianpursley** [dotnet/efcore#25413]
## File a bug
Using `.AsNoTracking()` is slower over multiple iterations of a simple query.

I'm using Sqlite if that makes a difference.

### Include your code

Here is the repo with database to reproduce the issue:
https://github.com/brianpursley/AsNoTrackingTest

Here is the code:
```C#
using System;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace AsNoTrackingTest
{
    class Program
    {
        static void Main(string[] args)
        {
            var useAsNoTracking = bool.Parse(args[0]);
            var iterations = int.Parse(args[1]);
            Console.WriteLine($"useAsNoTracking = {useAsNoTracking}, iterations = {iterations}");
            
            using var db = new NorthwindDbContext();
            for (var i = 0; i < iterations; i++)
            {
                if (useAsNoTracking)
                {
                    db.Customers.AsNoTracking().ToList();
                }
                else
                {
                    db.Customers.ToList();
                }
            }
        }
    }

    class NorthwindDbContext : DbContext
    {
        public DbSet<Customer> Customers { get; set; }
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        { 
            optionsBuilder.UseSqlite("Data Source=./Northwind_small.sqlite");
        }
    }

    [Table("Customer")]
    class Customer
    {
        public string Id { get; set; }
        public string CompanyName { get; set; }
        public string ContactName { get; set; }
        public string ContactTitle { get; set; }
        public string Address { get; set; }
        public string City { get; set; }
        public string Region { get; set; }
        public string PostalCode { get; set; }
        public string Country { get; set; }
        public string Phone { get; set; }
        public string Fax { get; set; }
    }
}
```

### Include stack traces
n/a

### Include verbose output

**Running without `.AsNoTracking()`**
10,000 iterations takes approx **5.1 seconds**:
```
$ time ./AsNoTrackingTest false 10000
useAsNoTracking = False, iterations = 10000

real	0m5.144s
user	0m4.822s
sys	0m0.433s
```

**Running with `.AsNoTracking()`**
10,000 iterations takes approx **9.6 seconds**:
```
$ time ./AsNoTrackingTest true 10000
useAsNoTracking = True, iterations = 10000

real	0m9.638s
user	0m9.316s
sys	0m0.445s
```

### Include provider and version information

EF Core version: **5.0.8**
Database provider: **Sqlite**
Target framework: **.NET 5.0.302**
Operating system: **Ubuntu 20.04.2 LTS (5.8.0-63-generic)**
IDE: **n/a**

**Comment by roji** [comment]
I could not reproduce your results, see my benchmark code below. As you did not provide the actual data being fetched, I used [the Northwind database used in the EF Core tests](https://github.com/dotnet/efcore/blob/main/test/EFCore.Sqlite.FunctionalTests/northwind.db), measuring both fetching the entire Customers table, and one single row.

Your benchmark has some issues which may skew results: there is no accounting for warm-up, and the number of iterations seems arbitrarily-chosen. Always consider using [BenchmarkDotNet](https://benchmarkdotnet.org/index.html) for benchmarks - it generally produces reliable, reproducible results. If you're still seeing that AsNoTracking is slower, please tweak the below code to make this show up.

BenchmarkDotNet=v0.13.0, OS=ubuntu 20.10
Intel Xeon W-2133 CPU 3.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.100-preview.7.21377.35
  [Host]     : .NET 6.0.0 (6.0.21.37712), X64 RyuJIT
  DefaultJob : .NET 6.0.0 (6.0.21.37712), X64 RyuJIT


|             Method |       Mean |   Error |  StdDev |
|------------------- |-----------:|--------:|--------:|
|   Tracking_AllRows | 1,426.2 us | 3.48 us | 3.25 us |
| NoTracking_AllRows | 1,231.8 us | 3.42 us | 2.85 us |
|    Tracking_OneRow |   876.2 us | 2.90 us | 2.57 us |
|  NoTracking_OneRow |   865.3 us | 2.42 us | 2.26 us |

<details>
<summary>Benchmark code</summary>

```c#
public class Program
{
    [Benchmark]
    public List<Customer> Tracking_AllRows()
    {
        using var db = new NorthwindDbContext();
        return db.Customers.ToList();
    }

    [Benchmark]
    public List<Customer> NoTracking_AllRows()
    {
        using var db = new NorthwindDbContext();
        return db.Customers.AsNoTracking().ToList();
    }

    [Benchmark]
    public List<Customer> Tracking_OneRow()
    {
        using var db = new NorthwindDbContext();
        return db.Customers.Take(1).ToList();
    }

    [Benchmark]
    public List<Customer> NoTracking_OneRow()
    {
        using var db = new NorthwindDbContext();
        return db.Customers.Take(1).AsNoTracking().ToList();
    }

    static void Main(string[] args)
        => BenchmarkRunner.Run<Program>();
}

class NorthwindDbContext : DbContext
{
    public DbSet<Customer> Customers { get; set; }
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.UseSqlite("Data Source=/home/roji/projects/efcore/test/EFCore.Sqlite.FunctionalTests/northwind.db");
    }
}

[Table("Customers")]
public class Customer
{
    public string CustomerId { get; set; }
    public string CompanyName { get; set; }
    public string ContactName { get; set; }
    public string ContactTitle { get; set; }
    public string Address { get; set; }
    public string City { get; set; }
    public string Region { get; set; }
    public string PostalCode { get; set; }
    public string Country { get; set; }
    public string Phone { get; set; }
    public string Fax { get; set; }
}
```

**Comment by ajcvickers** [comment]
@brianpursley Note that there is no guarantee that AsNoTracking will be faster. I haven't tried this for a while, but historically it has always been possible to construct cases where tracking is faster by making the materialization of objects dominate the query time. A repeated tracking query will then be faster because it only materializes the objects once, while a no-tracking query creates new instances every time.

**Comment by brianpursley** [comment]
Thanks @roji for taking a look at this.

> As you did not provide the actual data being fetched

I did provide it in the repo I linked, but that's OK.  I will change to use the database you linked.

> Always consider using BenchmarkDotNet for benchmarks

Thanks, I didn't know about this.  I updated my test to use this tool.  New code below:

```
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using Microsoft.EntityFrameworkCore;

namespace AsNoTrackingTest
{
    public class Program
    {
        
        [Benchmark]
        public void Tracking_OneRow()
        {
            using var db = new NorthwindDbContext();
            db.Customers.Take(1).ToList();
        }
        
        [Benchmark]
        public void NoTracking_OneRow()
        {
            using var db = new NorthwindDbContext();
            db.Customers.Take(1).AsNoTracking().ToList();
        }
        
        [Benchmark]
        public void Tracking_OneRow_Twice()
        {
            using var db = new NorthwindDbContext();
            db.Customers.Take(1).ToList();
            db.Customers.Take(1).ToList();
        }
        
        [Benchmark]
        public void NoTracking_OneRow_Twice()
        {
            using var db = new NorthwindDbContext();
            db.Customers.Take(1).AsNoTracking().ToList();
            db.Customers.Take(1).AsNoTracking().ToList();
        }
        
        [Benchmark]
        public void Tracking_AllRows()
        {
            using var db = new NorthwindDbContext();
            db.Customers.ToList();
        }
        
        [Benchmark]
        public void NoTracking_AllRows()
        {
            using var db = new NorthwindDbContext();
            db.Customers.AsNoTracking().ToList();
        }
        
        [Benchmark]
        public void Tracking_AllRows_Twice()
        {
            using var db = new NorthwindDbContext();
            db.Customers.ToList();
            db.Customers.ToList();
        }
        
        [Benchmark]
        public void NoTracking_AllRows_Twice()
        {
            using var db = new NorthwindDbContext();
            db.Customers.AsNoTracking().ToList();
            db.Customers.AsNoTracking().ToList();
        }

        [Benchmark]
        public void Tracking_Disjoint()
        {
            using var db = new NorthwindDbContext();
            db.Customers.Take(40).ToList();
            db.Customers.Skip(40).ToList();
        }
        
        [Benchmark]
        public void NoTracking_Disjoint()
        {
            using var db = new NorthwindDbContext();
            db.Customers.Take(40).AsNoTracking().ToList();
            db.Customers.Skip(40).AsNoTracking().ToList();
        }

        static void Main(string[] args)
            => BenchmarkRunner.Run<Program>();
    }

    class NorthwindDbContext : DbContext
    {
        public DbSet<Customer> Customers { get; set; }
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        { 
            optionsBuilder.UseSqlite("Data Source=northwind.db");
        }
    }

    public class Customer
    {
        public string CustomerId { get; set; }
        public string CompanyName { get; set; }
        public string ContactName { get; set; }
        public string ContactTitle { get; set; }
        public string Address { get; set; }
        public string City { get; set; }
        public string Region { get; set; }
        public string PostalCode { get; set; }
        public string Country { get; set; }
        public string Phone { get; set; }
        public string Fax { get; set; }
    }
}
```

BenchmarkDotNet=v0.13.0, OS=ubuntu 20.04
Intel Core i7-10710U CPU 1.10GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=5.0.302
  [Host]     : .NET 5.0.8 (5.0.821.31504), X64 RyuJIT
  DefaultJob : .NET 5.0.8 (5.0.821.31504), X64 RyuJIT


|                   Method |       Mean |    Error |   StdDev |
|------------------------- |-----------:|---------:|---------:|
|          Tracking_OneRow |   974.6 us | 13.60 us | 12.73 us |
|        NoTracking_OneRow |   962.6 us | 11.81 us | 10.47 us |
|    Tracking_OneRow_Twice | 1,839.1 us | 14.55 us | 12.15 us |
|  NoTracking_OneRow_Twice | 1,847.6 us | 27.27 us | 22.77 us |
|         Tracking_AllRows | 1,668.9 us | 14.88 us | 12.42 us |
|       NoTracking_AllRows | 1,431.1 us | 12.65 us | 10.56 us |
|   Tracking_AllRows_Twice | 2,632.7 us | 30.24 us | 26.81 us |
| NoTracking_AllRows_Twice | 2,830.8 us | 51.95 us | 46.06 us |
|        Tracking_Disjoint | 2,611.6 us | 34.35 us | 30.45 us |
|      NoTracking_Disjoint | 2,367.2 us | 40.59 us | 35.98 us |

**NoTracking_OneRow_Twice** and **NoTracking_AllRows_Twice** are slower than the Tracking versions.

---

> Note that there is no guarantee that AsNoTracking will be faster. 
 
@ajcvickers Well maybe its not a "guarantee" but the [Microsoft docs say they are "quicker to execute"] (https://docs.microsoft.com/en-us/ef/core/querying/tracking#no-tracking-queries). Probably what I'm doing isn't a real-world scenario and would be considered an edge case because why would you intentionally retrieve the same data twice.

I think if you look at my "Disjoint" test cases though, it shows that when retrieving **different** rows, NoTracking is (as expected) going to be faster.

I'll go ahead and close this issue.  I think it can be explained by what @ajcvickers said "A repeated tracking query will then be faster because it only materializes the objects once, while a no-tracking query creates new instances every time".

Feel free to re-open if you think otherwise.

Thanks, both of you.

**Comment by roji** [comment]
>> As you did not provide the actual data being fetched
>
> I did provide it in the repo I linked, but that's OK. I will change to use the database you linked.

Apologies, I missed that.

Apart from that, yeah, the "Twice" scenarios probably show exactly what @ajcvickers mentioned above: since all entity instances are already tracked from the 1st load, the 2nd load is considerably faster. This manages to offset the general cost of change tracking, which is an additional layer in addition to what happens in NoTracking, and which has its own overhead cost (which you can see when comparnig the non-Twice scenarios).

Bottom line, results seem pretty much in line with what we'd expect. If you run into anything that looks odd, don't hesitate to post back!
